app/models/experimentation/treatment_group_membership.rb

```diff
@@ -36,9 +36,40 @@ def record_notification_result(message, delivery_result)
       save!
     end
 
+    def record_visit_details(blood_pressure:, blood_sugar:, prescription_drug:)
+      visits = [blood_pressure, blood_sugar, prescription_drug].compact
+      return if visits.blank?
+
+      earliest_visit = visits.min_by { |visit| recorded_or_device_created_at(visit) }
+      visit_date = recorded_or_device_created_at(earliest_visit)
+      visit_facility = earliest_visit.facility
+
+      update!(
+        visit_blood_pressure_id: blood_pressure&.id,
+        visit_blood_sugar_id: blood_sugar&.id,
+        visit_prescription_drug_created: prescription_drug.present?,
+        visit_date: visit_date,
+        visit_facility_id: visit_facility.id,
+        visit_facility_name: visit_facility.name,
+        visit_facility_type: visit_facility.facility_type,
+        visit_facility_block: visit_facility.block,
+        visit_facility_district: visit_facility.district,
+        visit_facility_state: visit_facility.state,
+        status: :visited,
+        status_reason: :visit_recorded,
+        days_to_visit: (visit_date.to_date - experiment_inclusion_date.to_date).to_i
+      )
+    end
+
     private
 
+    def recorded_or_device_created_at(visit)
+      visit.recorded_at.presence || visit.device_created_at
```

Comment:
I remember creating aliases for `recorded_at` in entities that didn't have it. Recheck if you really need to do this.

Reply:
Ah great https://github.com/simpledotorg/simple-server/blob/7d972f2550ad0ec9d4407d2e9d73ab4ccc273539/app/models/prescription_drug.rb#L24. Thanks 🙌🏼